kernel/src/hil/spi.rs

```diff
@@ -147,3 +147,67 @@ pub trait SpiMasterDevice {
     fn get_phase(&self) -> ClockPhase;
     fn get_rate(&self) -> u32;
 }
+
+pub trait SpiSlaveClient {
+    // This is called whenever the slave is selected by the master
+    fn chip_selected(&self);
+
+    // This is called as a DMA interrupt when a transfer has completed
+    fn read_write_done(&self,
+                       write_buffer: Option<&'static mut [u8]>,
+                       read_buffer: Option<&'static mut [u8]>,
+                       len: usize);
+}
+
+pub trait SpiSlave {
+    fn init(&self);
+    fn has_client(&self) -> bool;
+    // return if there is a client.
+
+    fn set_client(&self, client: Option<&'static SpiSlaveClient>);
+
+    // buffer level or byte-level?
+    // fn set_transfer_mode(&self, transfor_mode: TransferMode);
+    // fn get_transfer_mode(&self) -> TransferMode;
+
+    fn set_write_byte(&self, write_byte: u8);
+    fn read_write_bytes(&self,
+                        write_buffer: Option<&'static mut [u8]>,
+                        read_buffer: Option<&'static mut [u8]>,
+                        len: usize)
+                        -> ReturnCode;
+    // fn write_byte(&self, val: u8);
+    // fn read_byte(&self) -> u8;
+    // fn read_write_byte(&self, val: u8) -> u8;
+
+    fn read_write_byte(&self, write_byte: u8) -> u8;
```

Comment:
Alright, sounds good. `SpiMaster` has the same synchronous interface, although it seems largely deprecated. Should we also remove it from the master interface?